### PR TITLE
feat: add collectible plugin

### LINKED ISSUE #2651: [Demand] Introduce Collectible Plugin
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

For now, the NFT plugin supports recognize NFT assets by parsing an OpenSea link. We may extend this feature to support various links from different platforms like Rarible and SuperRare. Also, we may provide the user to share those NFT assets listed in the collectible list as a post with a specific version of payload.  

If the plugin detected an NTF asset.

For an NFT asset owned by others. It may let the user
  + view the NFT asset info (price, traits, trade history, etc.) from some backend data sources.
  + buy an NFT asset with an Ethereum wallet to be connected.
  + offer price under an NFT asset.
  + accept the offers created by other users.

For an owned NFT asset. It may let the user
  + transfer it to another address.
  + sell it by creating an auction or bidding.

## Notes

+ [UI](https://www.figma.com/file/gVkQ67y285b4FXVV1KPThN/Twitter?node-id=8896:0)
+ [OpenSea JS SDK](https://github.com/ProjectOpenSea/opensea-js)
+ [Demo Project](https://ships-log.herokuapp.com/)

### LINKED ISSUE #2684: [Demand]  Read collectible token from link or customized payload
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Collectible)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

For any link represents a collectible on Twitter. It's reasonable to read token information from it. Here is a list example links from digital asset marketplaces and crypto projects.

| Provider or Porject | Example | Contract Address | Token ID |
| -------- | ------- | ---------------- | ---------|
| OpenSea  | https://opensea.io/assets/0x51e613727fdd2e0b91b51c3e5427e9440a7957e4/12107227 | `0x51e613727fdd2e0b91b51c3e5427e9440a7957e4` | `12107227` |
| Rarible | https://rarible.com/token/0xd07dc4262bcdbf85190c01c996b4c06a461d2430:42686 | `0xd07dc4262bcdbf85190c01c996b4c06a461d2430` | `42686` |
| SuperRare | https://superrare.co/artwork/hello-moon-1631 | `0x41a322b28d0ff354040e2cbc676f0320d8c8850d` | `1631` |
| NBA Top Shot | https://nbatopshot.com/listings/p2p/208ae30a-a4fe-42d4-9e51-e6fd1ad2a7a9+e5643680-9741-45db-8fbd-fc00dc107022 | \- | \- | 
| CryptoPunks | https://www.larvalabs.com/cryptopunks/details/8951 | `0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB` | `8951` |
| CryptoKitty | https://www.cryptokitties.co/kitty/1971632 |`0x06012c8cf97BEaD5deAe237070F9587f8E7A266d` | `1971632` |
| ENS | https://app.ens.domains/name/test.eth | \- | \- |

## Part 2: Problem to Solve

Read collectible `token_address` and `token_id` by parsing above links.

## Part 3: Expected Behavior

For any tweet contains any link in the above table. The Mask Network should present a Collectible card under that tweet.

## Notes

### LINKED ISSUE #2685: [Demand]  Implement Offers Tab (Collectible)
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Collectible)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Render price offers as a table in the Offers tab.

image

### Accept Offer

+ If no wallet is connected, then it presents a `Connect Wallet` button.
+ If the connected wallet is the collectible owner and at least one valid price offer exists. Each offer in the table can be selected and render an `Accept Offer` button if selected any of them.
+ Click the button will open the `Accept an Offer` dialog.

image

### Make Offer

+ If no wallet is connected, then it presents a `Connect Wallet` button.
+ If a collectible is valid and listed for accepting offers and the connected wallet is not the collectible owner, render a `Make Offer` button.
+ Click the button will open the `Make an Offer` dialog.

image

### Place a Bid

+ If no wallet is connected, then it presents a `Connect Wallet` button.
+ If a collectible is valid and listed for accepting offers and the connected wallet is not the collectible owner, render a `Place a Bid` button.
+ Click the button will open the `Place a Bid` dialog.

image

### Cancel Offer

+ If no wallet is connected, then it presents a `Connect Wallet` button.
+ If the connected wallet is the Offer's creator, it will render a `Cancel` button in each table row.

### Example tweet

image

https://twitter.com/suji_yan/status/1375167910633644032?s=20

## Notes

### LINKED ISSUE #2686: [Demand]  Implement Listing Tab (Collectible)
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Collectible)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Render orders as a table in the Listing tab.

### Create Order

+ If no wallet is connected, then it presents a `Connect Wallet` button.
+ If the connected wallet is the collectible owner, then render a `Sell` button.
+ Click the button will open the `Sell Collectible` dialog. The dialog will build with two tabs: `Set Price` and `Highest Bid`.

image

### Cancel Order

+ If no wallet is connected, then it presents a Connect Wallet button.
+ If the connected wallet Order's creator, it will render a `Cancel` button in each table row.

image

## Notes

### LINKED ISSUE #2687: [Demand]  Implement Details Tab (Collectible)
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Collectible)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Render all of the read-only information of a given collectible.

image

## Notes

### LINKED ISSUE #2689: [Demand]  Implement Article Tab (Collectible)
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Collectible)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

The article tab contains:

+ The collectible's hero image.

image

## Notes

### LINKED ISSUE #2695: [Demand] Implement Trade History Tab (Collectible)
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Collectible)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Render trade history as a table in the History tab.

image

### PR DESCRIPTION
**Pull request by guanbinrui**
## Related Issues

+ close #2651
+ close #2684 Read collectible token information
+ close #2695 Implement History Tab
+ close #2689 Implement Article Tab
+ close #2687 Implement Details Tab
+ close #2686 Implement Listing Tab
+ close #2685 Implement Offers Tab

## 

- [x] Read Collectible token infomation
- [x] Article Tab
- [x] Details Tab
- [x] Listing Tab
- [x] Hisotry Tab
- [x] Offers Tab

## Figma

https://www.figma.com/file/gVkQ67y285b4FXVV1KPThN/Twitter?node-id=8896:0

## Screenshots

image
image
image
image
image

## Notes

[OpenSea JS SDK](https://github.com/ProjectOpenSea/opensea-js)
[OpenSea's Testnet](https://testnets.opensea.io)

### CODE REVIEW

File: packages/maskbook/src/plugins/Collectible/schema.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "type": "object",
+    "additionalProperties": true,
+    "definitions": {}
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Collectible/schema.json:4]: 🤔?

File: packages/maskbook/package.json [hunk 2]
```diff
@@ -58,6 +58,7 @@
     "crypto-browserify": "^3.12.0",
     "d3": "^5.16.0",
     "date-fns": "^2.16.1",
+    "dayjs": "^1.10.4",
```
Comment (Jack-Works) [inline packages/maskbook/package.json:61]: Please don't introduce too many date APIs. We already have date-fns

**Comment by septs** [comment]
@nuanyang233 plz solve all warning.

**Comment by nuanyang233** [comment]
> @nuanyang233 plz solve all warning.

There no warning in ci. Where are you referring the warning

**Comment by septs** [comment]
https://github.com/DimensionDev/Maskbook/actions/runs/732648068

File: packages/maskbook/src/plugins/Collectible/UI/HistoryTab/index.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                        </TableCell>
+                    </TableRow>
+                </TableFooter>
+            </Table>
+        )
+
+    if (!events.value || events.error)
+        return (
+            <Box className={classes.empty}>
+                <Typography color="textSecondary">No History</Typography>
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Collectible/UI/HistoryTab/index.tsx:105]: Lost in `message.json`.

File: packages/maskbook/src/plugins/Collectible/UI/OrderRow.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                                        alt={order.paymentTokenContract?.symbol}
+                                    />
+                                )}
+                            </Link>
+                            {order.unitPrice}
+                        </Typography>
+                    </TableCell>
+                    <TableCell>
+                        <Typography>
+                            {new BigNumber(order.expirationTime).toString() !== '0' &&
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Collectible/UI/OrderRow.tsx:119]:
```suggestion
                            {!(new BigNumber(order.expirationTime).isZero()) &&
```

File: packages/maskbook/src/plugins/Collectible/UI/TokenTab.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                <Typography variant="h6">{t('plugin_collectible_chain_info')}</Typography>
+
+                <Box className={classes.chain_row}>
+                    <Typography>{t('plugin_collectible_contract_address')}</Typography>
+                    <Link
+                        href={`https://etherscan.io/address/${token?.contractAddress ?? ''}`}
+                        target="_blank"
+                        rel="noopener noreferrer">
+                        <Typography>
+                            {token?.contractAddress.slice(0, 6)}...{token?.contractAddress.slice(-6)}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Collectible/UI/TokenTab.tsx:119]: `formatEthereumAddress`

File: packages/maskbook/src/plugins/Collectible/UI/TokenTab.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                </Box>
+            ) : null}
+
+            <Box className={classes.container}>
+                <Typography variant="h6">{t('plugin_collectible_chain_info')}</Typography>
+
+                <Box className={classes.chain_row}>
+                    <Typography>{t('plugin_collectible_contract_address')}</Typography>
+                    <Link
+                        href={`https://etherscan.io/address/${token?.contractAddress ?? ''}`}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Collectible/UI/TokenTab.tsx:115]: Add one more pipe in `packages/maskbook/src/web3/pipes.ts`.

File: packages/maskbook/src/plugins/Collectible/apis/opensea.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
...data\n          id\n        }\n        seller {\n          ...AccountLink_data\n          id\n        }\n        toAccount {\n          ...AccountLink_data\n          id\n        }\n        winnerAccount {\n          ...AccountLink_data\n          id\n        }\n        ...EventTimestamp_data\n        id\n        __typename\n      }\n      cursor\n    }\n    pageInfo {\n      endCursor\n      hasNextPage\n    }\n  }\n}\n\nfragment EventTimestamp_data on AssetEventType {\n  eventTimestamp\n  transaction {\n    blockExplorerLink\n    id\n  }\n}\n\nfragment Price_data on AssetType {\n  decimals\n  imageUrl\n  symbol\n  usdSpotPrice\n  assetContract {\n    blockExplorerLink\n    id\n  }\n}\n\nfragment ProfileImage_data on AccountType {\n  imageUrl\n  address\n  chain {\n    identifier\n    id\n  }\n}\n\nfragment asset_url on AssetType {\n  assetContract {\n    account {\n      address\n      chain {\n        identifier\n        id\n      }\n      id\n    }\n    id\n  }\n  tokenId\n}\n\nfragment quantity_data on AssetQuantityType {\n  asset {\n    decimals\n    id\n  }\n  quantity\n}\n\nfragment wallet_accountKey on AccountType {\n  address\n  chain {\n    identifier\n    id\n  }\n}\n',
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Collectible/apis/opensea.ts:56]: Woo, place this big chunk of text into a single file?

File: packages/maskbook/src/plugins/Collectible/UI/Collectible.tsx [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                                    </Typography>
+                                </Box>
+                            ) : null}
+
+                            {currentPrice ? (
+                                <Box display="flex" alignItems="center" sx={{ marginTop: 1 }}>
+                                    <Typography
+                                        className={classes.description}
+                                        component="span"
+                                        dangerouslySetInnerHTML={{
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Collectible/UI/Collectible.tsx:162]: https://react.i18next.com/latest/trans-component

File: packages/maskbook/src/plugins/Collectible/apis/opensea.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
+import { OpenSeaPort } from 'opensea-js'
+import type { OrderSide, WyvernSchemaName } from 'opensea-js/lib/types'
+import stringify from 'json-stable-stringify'
+import { getChainId } from '../../../extension/background-script/EthereumService'
+import { resolveOpenSeaNetwork } from '../pipes'
+import { OpenSeaBaseURL, OpenSeaGraphQLURL, ReferrerAddress } from '../constants'
+import { Flags } from '../../../utils/flags'
+import type { OpenSeaAssetEventResponse, OpenSeaResponse } from '../UI/types'
+
+const apiKey = 'c38fe2446ee34f919436c32db480a2e3'
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Collectible/apis/opensea.ts:10]: Is it OK?

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/Collectible/apis/opensea.ts (hunk 9 above)
Comment (guanbinrui) [reply packages/maskbook/src/plugins/Collectible/apis/opensea.ts:10]: We have many hardcoded API keys that, if added as CI environment variables, will be much better.

File: packages/maskbook/src/plugins/Collectible/queries/OpenSea.ts [hunk 10]
```diff
@@ -0,0 +1,2 @@
...
..._data\n          id\n        }\n        seller {\n          ...AccountLink_data\n          id\n        }\n        toAccount {\n          ...AccountLink_data\n          id\n        }\n        winnerAccount {\n          ...AccountLink_data\n          id\n        }\n        ...EventTimestamp_data\n        id\n        __typename\n      }\n      cursor\n    }\n    pageInfo {\n      endCursor\n      hasNextPage\n    }\n  }\n}\n\nfragment EventTimestamp_data on AssetEventType {\n  eventTimestamp\n  transaction {\n    blockExplorerLink\n    id\n  }\n}\n\nfragment Price_data on AssetType {\n  decimals\n  imageUrl\n  symbol\n  usdSpotPrice\n  assetContract {\n    blockExplorerLink\n    id\n  }\n}\n\nfragment ProfileImage_data on AccountType {\n  imageUrl\n  address\n  chain {\n    identifier\n    id\n  }\n}\n\nfragment asset_url on AssetType {\n  assetContract {\n    account {\n      address\n      chain {\n        identifier\n        id\n      }\n      id\n    }\n    id\n  }\n  tokenId\n}\n\nfragment quantity_data on AssetQuantityType {\n  asset {\n    decimals\n    id\n  }\n  quantity\n}\n\nfragment wallet_accountKey on AccountType {\n  address\n  chain {\n    identifier\n    id\n  }\n}\n'
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Collectible/queries/OpenSea.ts:2]: Please use ES6 multiple-line string

**Review by Jack-Works — APPROVED** [review]

**Comment by Jack-Works** [comment]
Conflicting files
pnpm-lock.yaml

File: packages/maskbook/src/plugins/PluginUI.ts [hunk 11]
```diff
@@ -26,12 +27,13 @@ sideEffect.then(() => {
     plugins.add(RedPacketPluginDefine)
     plugins.add(FileServicePluginDefine)
     plugins.add(ITO_PluginDefine)
-    plugins.add(NFT_PluginsDefine)
+    // plugins.add(NFTPluginsDefine)
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/PluginUI.ts:30]: So this is going to remove NFT plugin? cc @guanbinrui

Comment (guanbinrui) [reply packages/maskbook/src/plugins/PluginUI.ts:30]: Will add back if the plugin on/off function is ready.

Comment (Jack-Works) [reply packages/maskbook/src/plugins/PluginUI.ts:30]: why close it for now?

**Review by Jack-Works — DISMISSED** [review]

**Comment by Jack-Works** [comment]
so many merge conflicts

**Review by septs — CHANGES_REQUESTED** [review]
1. clean `console.*`
2. use https://aka.pw/svgomg reduce svg icon

**Comment by guanbinrui** [comment]
> 1. clean `console.*`
> 2. use https://aka.pw/svgomg reduce svg icon

ALL DONE!

**Review by septs — APPROVED** [review]

**Comment by Jack-Works** [comment]
hold on. still reviewing

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**